chunk/chunk_store.go

```diff
@@ -243,49 +243,128 @@ outer:
 	return filteredChunks, nil
 }
 
-func (c *Store) lookupMatchers(ctx context.Context, from, through model.Time, matchers []*metric.LabelMatcher) ([]Chunk, error) {
-	metricName, matchers, err := util.ExtractMetricNameFromMatchers(matchers)
+func (c *Store) lookupChunksByMatchers(ctx context.Context, from, through model.Time, matchers []*metric.LabelMatcher) ([]Chunk, error) {
+	metricNameMatcher, matchers, ok := util.ExtractMetricNameMatcherFromMatchers(matchers)
+
+	// Only lookup by metric name if the matcher is of type equal, otherwise we
+	// have to fetch chunks for all metric names as other metric names could match.
+	if ok && metricNameMatcher.Type == metric.Equal {
+		return c.lookupChunksByMetricName(ctx, from, through, matchers, metricNameMatcher.Value)
+	}
+
+	userID, err := user.Extract(ctx)
 	if err != nil {
 		return nil, err
 	}
 
+	// If there is no metric name, we want return chunks for all metric names
+	metricNameQueries, err := c.schema.GetReadQueries(from, through, userID)
+	if err != nil {
+		return nil, err
+	}
+	metricNameEntries, err := c.lookupEntriesByQueries(ctx, metricNameQueries)
+	if err != nil {
+		return nil, err
+	}
+
+	incomingChunkSets := make(chan ByKey)
+	incomingErrors := make(chan error)
+	skippedMetricNames := 0
+
+	for _, metricNameEntry := range metricNameEntries {
+		metricName, err := parseMetricNameRangeValue(metricNameEntry.RangeValue, metricNameEntry.Value)
+		if err != nil {
+			return nil, err
+		}
+
+		// If there is a metric name matcher, however we are fetching all metric name
+		// chunks, skip metric names which don't match the metric name matcher.
```

Comment:
This doesn't make sense grammatically. I *think* it means…

```golang
// We are fetching all metric name chunks, but we only want metric names that match metricNameMatcher.
```

Is that right?